Add: text on Hatch backend compatibility (#228)
pyOpenSci/python-package-guide: package-structure-code/complex-python-package-builds.md

1. **Pure-python packages:** these are packages that only rely on Python to function. Building a pure Python package is simpler. As such, you can chose a tool below that has the features that you want and be done with your decision!

> 2. **Python packages with non-Python extensions:** These packages have additional components called extensions written in other languages (such as C or C++). If you have a package with non-python extensions, then you need to select a build back-end tool that allows you to add additional build steps needed to compile your extension code. Further, if you wish to use a front-end tool to support your workflow, you will need to select a tool that supports additional build setups. We suggest that you chose build tool that supports custom build steps like Hatch.

Reviewer: ```suggestion
2. **Python packages with non-Python extensions:** These packages have additional components called extensions written in other languages (such as C or C++). If you have a package with non-Python extensions, then you need to select a build backend tool that allows additional build steps needed to compile your extension code. Further, if you wish to use a front-end tool (such as JavaScript) to support your workflow, you will need to select a tool that supports additional build setups. We suggest that you chose build tool that supports custom build steps like Hatch.
```

Reviewer: A small nit (@lwasser): what is our preferred spelling of front-end and back-end?

Author: FWIW I was following the style in this file which was consistently `-end`.

Also not sure why you added JavaScript? Can JS build python projects and is this common? I was referring to frontends like `build` and `pip`.

Reviewer 2: i think i tend to use the - as @ucodery this would be another good place to have that vale style checker. as many including Jeremiah have pointed out I often forget which style we prefer.

Author: I think I've made all the requested changes, except everything still uses"-end".

(except that one time `build-backend` is used. A reason for a style change @lwasser ? "build-back-end" is probably not what we want 🤣 )

Reviewer: Ah @ucodery ...I was thinking about JupyterLab and how we distribute JS code and Python code. All looks good to me.

Reviewer 2: @ucodery i think build-backend is good ... i'm sorry i think i missed that it was spelled build-back-end 😆 agreed that is not what we want. let me actually look at the text here ...

Reviewer 2: @willingc  @ucodery i left another note inline. i have no preference. i say go with what is simplest and i'll add it to vale locally to remind myself not to go rogue 😆 or we do a slack poll. i really have no opinion here but agree consistency is ideal.

3. **Python packages that have extensions written in different languages (e.g. Fortran and C++) or that have non Python dependencies that are difficult to install (e.g. GDAL)** These packages often have complex build steps (more complex than a package with just a few C extensions for instance). As such, these packages require tools such as [scikit-build](https://scikit-build.readthedocs.io/en/latest/)
   or [meson-python](https://mesonbuild.com/Python-module.html) to build. NOTE: you can use meson-python with PDM.

## Mixing front-end and back-end projects

It is sometimes necessary or desirable to use a build front-end with an alternative build-backend.
This is because some front-ends do not have a default back-end (`build`), and this choice is placed on the maintainer.
Other back-ends (`hatch`) have a preferred back-end (`hatchling`) but allow the maintainer to migrate to another, while
some back-ends (`poetry`) only work with a single back-end (`poetry-core`). Refer to (#python-package-build-tools) for
more information about front-end and back-end compatibility.

> In this packaging guide we recommend using `hatch` along with it's preferred back-end `hatchling`. One reason to use

Reviewer: ```suggestion
In this packaging guide we recommend using `hatch` along with it's preferred back-end `hatchling`. While this will be suitable for most packages, an alternate back-end may be used with Hatch if needed when creating an extension module.
```

> hatch with an alternate back-end is for creating an extension module. A Python extension module is one that is make up,

Reviewer: ```suggestion
hatch with an alternate back-end is for creating an extension module. A Python extension module is one that is made up,
```

> hatch with an alternate back-end is for creating an extension module. A Python extension module is one that is make up,

Reviewer: ```suggestion
A Python extension module is one that is make up,
```

Author: I think you are suggesting to remove the entire sentence, starting with "One reason"?

> either in part or entirely, of compiled code. In this case the back-end chosen (such as `meson-python`) must know how to

Reviewer 2: oooh this is good @ucodery have you actually tested this by chance? i know a while back i tested pdm but i haven't had a change to test out hatch with meson-python (which a lot of the scientific community is using). but this is a really excellent addition to this page!!

Author: I did no testing for this PR 🙈. I've built extension modules with setuptools or maturin but never coupled with hatch, and never used meson-Python for anything.

Author: Alright, I think I tested that everything works as expected. It's a little meta, but `meson-python` the project uses `mesonpy` as it's own backend. So I cloned the repo, set up a new virtualenv with only `hatch` in it, and built `meson-python` with no change to its pyproject.toml.

Reviewer 2: wow thank you!! you rock!

i'm curious - did your pyproject.toml look something like this:

https://github.com/pyOpenSci/examplePy/blob/main/example6_pdm_meson/pyproject.toml 

essentially that was my playing around with building using meson-python / mesonpy and pdm. please also note that i've only worked on pure python packages so my knowledge of complex builds is minimal!

Author: I used the pyproject.toml exactly as it came down from git - so it looked like this https://github.com/mesonbuild/meson-python/blob/main/pyproject.toml.
Using hatch only as a front-end means that it is in my python environment, but not necessarily defined in project code at all. (Not at all in this case. If the project was making use of hatch environments but not using hatchling, then the pyproject.toml would have a reference to both build tools).

> compile the extension language and bind it to Python but `hatchling` is only capable of building pure Python modules.

Reviewer: ```suggestion
compile the extension language and bind it to Python since `hatchling` is only capable of building pure Python modules.
```

> compile the extension language and bind it to Python but `hatchling` is only capable of building pure Python modules.

Reviewer 3: Not strictly true; there is no built-in functionality but you can build extension modules with plugins. For example, a major one is coming very soon: https://github.com/scikit-build/scikit-build-core/pull/637

1. **Pure-python packages:** these are packages that only rely on Python to function. Building a pure Python package is simpler. As such, you can chose a tool below that has the features that you want and be done with your decision!

> 2. **Python packages with non-Python extensions:** These packages have additional components called extensions written in other languages (such as C or C++). If you have a package with non-Python extensions, then you need to select a build back-end tool that allows additional build steps needed to compile your extension code. Further, if you wish to use a front-end tool to support your workflow, you will need to select a tool that supports additional build setups. We suggest that you chose build tool that supports custom build steps like Hatch.

Reviewer 2: @ucodery ok i think i see - i wrote back-end instead of backend originally. build-back-end just looks weird to me

and then i found [this](https://ell.stackexchange.com/questions/117383/what-is-the-correct-term-back-end-back-end-or-backend) 😆 

shall we use frontend and backend and then build-backend ?! we really need a style checker. i'm dangerous without some tool reminding me to be consistent.
